### PR TITLE
Issue #2861002 by mglaman: Coupon redemption form needs severe improv…

### PR DESCRIPTION
**Pull request by mglaman**
…ements and bug fixes

### CODE REVIEW

File: modules/promotion/src/Element/CouponRedemptionForm.php [hunk 1]
```diff
@@ -166,26 +214,39 @@ public static function validateForm(array &$element, FormStateInterface $form_st
...
-    $promotion = $promotion_storage->loadByCoupon($order_type, $order->getStore(), $coupon);
+    $order_storage = $entity_type_manager->getStorage('commerce_order');
+    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+    $order = $order_storage->load($element['#order_id']);
 
-    if (self::couponApplies($order, $promotion, $coupon)) {
-      $order->get('coupons')->appendItem($coupon);
+    if ($triggering_element['#name'] == 'remove_coupon') {
+      $order->get('coupons')->setValue([]);
       $order->save();
```
Comment (mglaman) [inline modules/promotion/src/Element/CouponRedemptionForm.php]: Order should only be saved in `rebuildForm` (rename it) when element button clicked.

File: modules/promotion/src/Element/CouponRedemptionForm.php [hunk 2]
```diff
@@ -166,26 +214,39 @@ public static function validateForm(array &$element, FormStateInterface $form_st
...
+
+        $coupon = $coupon_storage->loadByCode($coupon_code);
+        if ($coupon) {
+          /** @var \Drupal\commerce_order\Entity\OrderTypeInterface $order_type */
+          $order_type = $order_type_storage->load($order->bundle());
+          $promotion = $promotion_storage->loadByCoupon($order_type, $order->getStore(), $coupon);
+
+          if (self::couponApplies($order, $promotion, $coupon)) {
+            $order->get('coupons')->appendItem($coupon);
+            $order->save();
```
Comment (mglaman) [inline modules/promotion/src/Element/CouponRedemptionForm.php]: Order should only be saved in rebuildForm (rename it) when element button clicked.

File: modules/promotion/src/Element/CouponRedemptionForm.php [hunk 3]
```diff
@@ -166,26 +214,39 @@ public static function validateForm(array &$element, FormStateInterface $form_st
...
+            drupal_set_message(t('Coupon applied'));
```
Comment (mglaman) [inline modules/promotion/src/Element/CouponRedemptionForm.php]: Allow this text to be customized.

File: modules/promotion/src/Element/CouponRedemptionForm.php [hunk 4]
```diff
@@ -67,11 +72,14 @@ public function getInfo() {
...
+    if (empty($element['#order_id'])) {
+      throw new \InvalidArgumentException('The commerce_coupon_redemption_form element requires the #order_id property.');
     }
-    if (!$element['#order'] instanceof OrderInterface) {
-      throw new \InvalidArgumentException('The commerce_coupon_redemption_form #order property must be an order entity.');
+
+    $order_storage = \Drupal::entityTypeManager()->getStorage('commerce_order');
+    $order = $order_storage->load($element['#order_id']);
+    if (!$order instanceof OrderInterface) {
+      throw new \InvalidArgumentException('The commerce_coupon_redemption_form #order_id property must be an existing order entity.');
```
Comment (bojanz) [inline modules/promotion/src/Element/CouponRedemptionForm.php]: This argument message is misleading, need to be clear that we need the ID of an existing order, and not the .order itself. "The commerce_coupon_redemption #order_id must be a valid order ID." ?

File: modules/promotion/src/Element/CouponRedemptionForm.php [hunk 5]
```diff
@@ -100,12 +110,71 @@ public static function processForm(array $element, FormStateInterface $form_stat
       '#limit_validation_errors' => [
         array_merge($element['#parents'], ['code']),
       ],
+      '#submit' => [
+        [get_called_class(), 'elementButtonSubmit'],
```
Comment (bojanz) [inline modules/promotion/src/Element/CouponRedemptionForm.php]: Why not call elementButtonSubmit and elementButtonRemove closer to what they do? applyCoupon and removeCoupon?

File: modules/promotion/src/Element/CouponRedemptionForm.php [hunk 6]
```diff
@@ -92,21 +101,100 @@ public static function processForm(array $element, FormStateInterface $form_stat
...
       '#description' => $element['#description'],
+      '#access' => !$has_coupons,
     ];
     $element['apply'] = [
       '#type' => 'submit',
       '#value' => $element['#submit_title'],
       '#name' => 'apply_coupon',
       '#limit_validation_errors' => [
+        $element['#parents'],
         array_merge($element['#parents'], ['code']),
```
Comment (mglaman) [inline modules/promotion/src/Element/CouponRedemptionForm.php]: We need to pass `$element['#parents']` to ensure submit callback has proper values. Maybe we can kill second line.

**Pull request merged**